### DISCUSSION TITLE
Shell script auto-start hacking

### DISCUSSION THREAD

**Discussion by ghost** [termux/termux-app#2902] [Q&A] [answered]
Hi team,
I am looking for a way to rebuild the Termux App and add my shell script to run inside on start. Can you help me to find a starting point in the codebase for that? Where I can put my own shell script?

**Comment by Maxython** [comment]
`$PREFIX/bin/login`

**Reply by ghost** [reply]
I still don't see where is bin in the https://github.com/termux/termux-app. Looks like it is downloaded somewhere? 
`$ find . -name 'bin'`

**Reply by Maxython** [reply]
Are you looking for the `bin` directory in the home directory? He is not here! The `bin` directory is in termux's `$PREFIX` (`/data/data/com.termux/files/usr`).  
If you want to get into the bin directory then just run the following command.
```bash
cd $PREFIX/bin
```
I take it you're new to termux and the Linux kernel. I recommend looking into the system directories in termux.

**Reply by ghost** [reply]
I looked for the rootfs location, here is answer, https://github.com/termux/termux-app/discussions/2902#discussioncomment-3201833, thanks

**Comment by agnostic-apollo** [comment, marked as answer]
You need to either unzip bootstrap zips and add scripts in `/data/data/com.termux/files/usr/bin` and call them in `/data/data/com.termux/files/usr/etc/profile` and zip again, then upload new zips to new url and update url in `app/build.gradle` or update already download  bootstrap zips in `app/src/main/cpp`.

https://github.com/termux/termux-app/blob/211340781bcb0576ffe100fda2e0ed87d7333d53/app/build.gradle#L162

https://github.com/termux/termux-packages/releases

**Reply by ghost** [reply]
Thanks, it helps

**Reply by ghost** [reply]
What is the right way to build this .zip and keep my app transparent? Create a GitHub action that add my updated files to the upstream rootfs.zip and uploads it as an artifact? Any other ideas?

**Reply by agnostic-apollo** [reply]
That would work. Or support can be added in `build.gradle` to call `process_bootstrap_install.sh` script if an env variable is set. The script could be overridden in your repo with a git `ours` merge strategy or something and env variable exported from github settings.

**Reply by RalfWerner** [reply]
@VasilyKorzh the proposal of @agnostic-apollo is replacement of the bootstrap (`termux` environment).
If you just want to use **your own script** at the `termux` start, you can simply store this in `~/.bashrc` or if it already exists e.g. `s=/sdcard/Download/bashrc` outside of termux and you allow access (`termux-setup-storage`), this can also be done with ` ln -s $s .bashrc`.

**Reply by ghost** [reply]
@RalfWerner I need a simplest solution for the end user. User has to just install apk and it works. How I can put my own ~/.bashrc without changing an actual rootfs?

**Reply by agnostic-apollo** [reply]
If you are forking termux and making other changes, then custom bootstrap would be ideal. If not, then you can also just provide users a single line command that downloads a script with `curl` and executes it to download and setup all your other scripts.

**Reply by RalfWerner** [reply]
> If you are forking termux and making other changes, then custom bootstrap would be ideal

You are right, especially when the whole environment deviates from Termux. In the current case I use for my _bootstrap loop_ e.g:
`214: doLast {def b=new File(projectDir,"bootstrap-i686.zip"); if (b.exists()) return`
> How I can put my own ~/.bashrc without changing an actual **rootfs?**

The _bootstrap_ creates only `u=~/../usr` and some defaults in `~/.termux` (`~/.bashrc` missing at begin). Do you mean `$u` or `/` is **rootfs**?
If **end user** mean someone  who only install the APK and _never use_ the terminal command line, you could also use Termux:Widget to provide **scripts** to him in `~/.shortcuts`.
But then two APK has to install and `.shortcuts` has to copy to termux and widget has to placed on startscreen

**Reply by ghost** [reply]
> If end user mean someone who only install the APK and never use the terminal command line

Yes, this is a case
